### DISCUSSION TITLE
Dev pain: why do you use hardcoding and final keywords in symfony classes?

### DISCUSSION THREAD

**Discussion by prohalexey** [symfony/symfony#52303] [Q&A]
I'm trying new things in the symfony 6.3 and I found some "interesting code" in the HttpKernel component.

Let's look into the code:

**1. First problem - Hardcoding**

For example [QueryParameterValueResolver](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/QueryParameterValueResolver.php)

If we didn't found a parameter we throw an exception:
```
throw new NotFoundHttpException(sprintf('Missing query parameter "%s".', $name));
```
And if we could not validate parameters we throw another exception:
```
throw new NotFoundHttpException(sprintf('Invalid query parameter "%s".', $name));
```
But how did the author intend to get the name and type of the exception from this exception? preg_match?

**2. Second problem - using final in the public packages/components/shared code.**

I can't customize the exception type, I can't even extend the class, why can't I do that?
Why do I need to clone a class and maintain it completely If I just want to change a small part of the code or some behavior of a class ?

I think it's a bad idea to mark classed as final. And `new` statements in your classed must be configurable.

**Comment by ghost** [comment]
1. i don't understand what you mean. Could you reword?
2. extending is not for changing code because you don't like the original code/behavior or you are a lazy developer that just wants to reuse code that someone else wrote but you don't like a few lines.
Class might be final if it is an implementation of an interface or there is no reasons for someone to extend it (sometimes such a decision might be subjective - if you think different just open na issue and provide arguments why a given class should not be final).
In open source marking classes as final is really helpfull because it eases maintaining the code and keeping BC promise.
For your problem with exceptions I think there is no need not to make them final. Those are exceptions thown by let's say Component and exceptions from outside this component cannot be of this type (remember that extending class is also of type of the parent class)

**Reply by prohalexey** [reply]
1. NotFoundHttpException - Is too general here, because we pass `$name` to the exception message, but we cannot get this name then we catch an exception. The better way is throw NotFoundHttpExceptionWithMissingName with `$name` as parameter extending NotFoundHttpException, or create a factory method that could be extended

2. Like or dislike is not a reason. I just have to, because I use this framework as many other people. For now I can't use this class because it uses the same exception for different cases. And I can't change it in any way...but have to. One option is clone whole class, but the more clones, the more support needed.

**Comment by mbabker** [comment]
You picked one of the worst classes possible to create yet another-final-versus-open debate with.

1) In the request-response workflow of a Symfony application, the argument resolvers are generally going to be executed by framework-level code; outside of a `kernel.exception` event listener, your application code is really not going to see these exceptions directly
2) Yeah, this class is final.  It is a single method class that implements an interface.  So even if you wanted to customize its behavior (which may or may not break expectations in other parts of the code) through inheritance, you're re-implementing the entire class anyway.

> And `new` statements in your classed must be configurable.

Not all class instantiations need to be configurable.  This leads to overly complex and unmaintainable code.  Plus, some instantiations are just internal implementation details and have no reason for override/configuration (unless you really have a strong justification for why you should have control over instantiating a `\Reflector` for the enum portion of the match in that class).  Imagine if every class that throws exceptions required some form of factory service to be able to throw those exceptions.  See the issue?

**Reply by prohalexey** [reply]
1. You are wrong. I use `kernel.exception` to handle NotFoundHttpException. And it works. I want to show different responses instead of "not found" in case missing parameter and invalid parameter.
2. Yes, but it could have better class structure if the author cared more about the people who would use this class.

> See the issue?

Yes, but in this case we could have 2 variants. 
- more specific exceptions
- configurable exceptions

**Reply by mbabker** [reply]
In general, the argument resolvers are designed to throw HTTP exceptions to at least set the desired response code based on the API design, or they're throwing some non-HTTP exceptions generally indicating the app is misconfigured in some way.  WIthout adding a bunch of subclasses of those HTTP exception classes to capture these extra bits, parsing the message is your best option.  It's a tradeoff of keeping the code maintainable (is it really in anyone's benefit to have a `MissingQueryParameter extends NotFoundHttpException {}` and `InvalidQueryParameter extends NotFoundHttpException {}` here, or digging into the other argument resolvers, `InvalidDate extends NotFoundHttpException {}` or `InvalidUid extends NotFoundHttpException {}`?).

Define "better class structure" because this is entirely subjective.  Is the code really more maintainable if each chunk of logic is placed into a micro-method just so you can extend the class and customize the behavior?  I would argue that makes things worse overall.  And at what point through your customizations does it just stop making sense to extend the class and just replace it with your own copy (especially if you override the lone method required by the interface)?  At what point does this request turn into "loosen the typehint for the attribute because I want to replace it with my own but use my completely unrelated attribute class with this resolver", or "loosen the typehint because I don't want this to create a 404 response"?

```php
public function resolve(Request $request, ArgumentMetadata $argument): array
{
    if (!$attribute = $this->findAttribute($argument)) {
        return [];
    }

    $name = $this->getArgumentName($attribute, $argument);
    if (!$this->requestHasArgument($request, $name)) {
        if ($this->argumentIsNullableOrHasDefault($argument)) {
            return [];
        }

        throw $this->createNotFoundExceptionForMissingQueryParameter($request, $argument, $attribute, $name);
    }

    $value = $this->getValueFromRequest($request, $name);

    [ ... ]
}

protected function findAttribute(ArgumentMetadata $argument): ?MapQueryParameter
{
    return $argument->getAttributesOfType(MapQueryParameter::class)[0] ?? null;
}

protected function getArgumentName(MapQueryParameter $attribute, ArgumentMetadata $argument): string
{
    return $attribute->name ?? $argument->getName();
}

protected function requestHasArgument(Request $request, string $name): bool
{
    return $request->query->has($name);
}

protected function argumentIsNullableOrHasDefault(ArgumentMetadata $argument): bool
{
    return $argument->isNullable() || $argument->hasDefaultValue();
}

protected function createNotFoundExceptionForMissingQueryParameter(Request $request, ArgumentMetadata $argument, MapQueryParameter $attribute, string $name): NotFoundHttpException
{
    return new NotFoundHttpException(sprintf('Missing query parameter "%s".', $name));
}

protected function getValueFromRequest(Request $request, string $name): mixed
{
    return $request->query->all()[$name];
}
```

Side note, filling your argument with claims about the humans involved in writing and maintaining the code doesn't help things at all, "if the author cared more about the people who would use this class" potentially carries a lot more negative connotation than you may have intended.  You're better off sticking to the technical aspects of your arguments.

**Reply by prohalexey** [reply]
> Side note, filling your argument with claims about the humans involved in writing and maintaining the code doesn't help things at all, "if the author cared more about the people who would use this class" potentially carries a lot more negative connotation than you may have intended.

It's not about feelings)

I'm talking about public source code, which was created not for 1 person/company, but for society. If you, I, or anyone else (as the author) has this idea in mind, there are points in your code that help extend or change the basic behavior. This is because the main part here is the application built on top of the framework, not the framework itself. That's why it needs to be flexible as it can.

**Reply by stof** [reply]
Killing the ability of the framework to evolve will not help applications using the framework. Opening everything to inheritance allows applications to do everything they want, but it freezes the framework in time (as _any_ change in the framework because a BC break as those inheritance-based extension points being everywhere mean that each implementation detail becomes exposed in the API)

**Reply by prohalexey** [reply]
You have several ways to make it flexible:
- in some cases (like `... = new Class`) you could make protected micro method which returns that class or
- you could make it more specific or
- you could make it configurable with some config(including DI)

And all of that ways do not killing the ability of the framework to evolve. In case of BC you could even make new class with new signature for public/protected methods in new version of framework.

**Reply by stof** [reply]
Micro-methods kill the ability of refactoring the class as they expose very fine-grain implementation details so any change in the code might break the expectations of the child class.
And making configurable adds lots of complexity (that we then need to maintain). So we would need clear use cases and not adding extension points everywhere.

**Reply by prohalexey** [reply]
We could return interfaces instead of real classes. It will not kill the ability of refactoring the class.

**Comment by stof** [comment]
See https://symfony.com/doc/current/contributing/code/bc.html#changing-classes for the kind of changes that are allowed in a codebase that needs to respect backward compatibility (note that those rules are not specific to Symfony. They can be used by other open-source projects as they are related to the impact of such changes on child classes or callers).

See all the lines using the notes `7` and `8`. Those are the lines where the `Allowed: no` becomes a `yes` instead when the class or method is final. Those represent more than half of the lines where the general answer is `Allowed: no`, which means that final has a big impact on what kind of changes we can do in a new version.

And regarding the visibility, you can compare the answers for rows that are in the sections about private visibility (hint: all answers are `Allowed: yes`) vs other visibilities

This is the main reason why we don't make everything public or protected: this would basically forbid us to improve Symfony.
Thus, composition-based extension points are almost always more powerful than inheritance, and they can be preserved when refactoring. So we would rather introduce supported extension points than opening everything in the class (which is the spirit of the open/closed principle btw).

**Reply by prohalexey** [reply]
Well noted. I just want to say that in this case the code is like stone, you cannot understand what is happening(exceptions are the same), and you cannot change it in any way, neither inheritance nor composition. And I think it could be more flexible.
